File: crossecore-java/src/java_/AbstractTypeDeclarationBase.java
package java_;
import org.eclipse.emf.common.notify.Notification;
import org.eclipse.emf.common.notify.NotificationChain;
import org.eclipse.emf.common.notify.impl.NotificationImpl;
import org.eclipse.emf.ecore.InternalEObject;
import org.eclipse.emf.ecore.*;
import org.eclipse.emf.ecore.impl.ENotificationImpl;
public class AbstractTypeDeclarationBase 
extends BodyDeclarationImpl implements AbstractTypeDeclaration
{
	private com.crossecore.ocl.Set<TypeAccess> _usagesInTypeAccess;
	
	public com.crossecore.ocl.Set<TypeAccess> getUsagesInTypeAccess()
	{
		if(_usagesInTypeAccess==null){
			_usagesInTypeAccess = new com.crossecore.ocl.Set<TypeAccess>(TypeAccess.class, this, Java_PackageImpl.TYPE__USAGES_IN_TYPE_ACCESS, Java_PackageImpl.TYPE_ACCESS__TYPE);
		}
		return _usagesInTypeAccess;
	
	}
	private com.crossecore.ocl.OrderedSet<BodyDeclaration> _bodyDeclarations;
	
	public com.crossecore.ocl.OrderedSet<BodyDeclaration> getBodyDeclarations()
	{
		if(_bodyDeclarations==null){
			_bodyDeclarations = new com.crossecore.ocl.OrderedSet<BodyDeclaration>(BodyDeclaration.class, this, Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS, Java_PackageImpl.BODY_DECLARATION__ABSTRACT_TYPE_DECLARATION);
		}
		return _bodyDeclarations;
	
	}
	private com.crossecore.ocl.OrderedSet<Comment> _commentsBeforeBody;
	
	public com.crossecore.ocl.OrderedSet<Comment> getCommentsBeforeBody()
	{
		if(_commentsBeforeBody==null){
			_commentsBeforeBody = new com.crossecore.ocl.OrderedSet<Comment>(Comment.class, this, Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_BEFORE_BODY, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_BEFORE_BODY);
		}
		return _commentsBeforeBody;
	
	}
	private com.crossecore.ocl.OrderedSet<Comment> _commentsAfterBody;
	
	public com.crossecore.ocl.OrderedSet<Comment> getCommentsAfterBody()
	{
		if(_commentsAfterBody==null){
			_commentsAfterBody = new com.crossecore.ocl.OrderedSet<Comment>(Comment.class, this, Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_AFTER_BODY, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_AFTER_BODY);
		}
		return _commentsAfterBody;
	
	}
	public Package getPackage()
	{
		
			if (eContainerFeatureID() != Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE) return null;
			return (Package)eInternalContainer();
		
		
	}
	public void setPackage(Package value){
		if (value != eInternalContainer()) {
			NotificationChain msgs = null;
			if (eInternalContainer() != null){
				msgs = ((InternalEObject)eInternalContainer()).eInverseRemove(this, Java_PackageImpl.PACKAGE__OWNED_ELEMENTS, Package.class, msgs);
			}
			if (value != null){
				msgs = ((InternalEObject)value).eInverseAdd(this, Java_PackageImpl.PACKAGE__OWNED_ELEMENTS, Package.class, msgs);
			}
			msgs = basicSetPackage(value, msgs);
			if (msgs != null) {
				msgs.dispatch();
			}
		}
		else if (eNotificationRequired()){
			eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE , value, value));
		}
	}
	private com.crossecore.ocl.OrderedSet<TypeAccess> _superInterfaces;
	
	public com.crossecore.ocl.OrderedSet<TypeAccess> getSuperInterfaces()
	{
		if(_superInterfaces==null){
			_superInterfaces = new com.crossecore.ocl.OrderedSet<TypeAccess>(TypeAccess.class, this, Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__SUPER_INTERFACES, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__SUPER_INTERFACES);
		}
		return _superInterfaces;
	
	}

	
	@Override
	protected EClass eStaticClass() {
		return Java_PackageImpl.Literals.ABSTRACTTYPEDECLARATION;
	}
	
	@Override
	public NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS:
				return getBodyDeclarations().basicAdd((BodyDeclaration)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS:
				return getUsagesInTypeAccess().basicAdd((TypeAccess)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE:
				if (eInternalContainer() != null) {
					msgs = eBasicRemoveFromContainer(msgs);
				}
				return basicSetPackage((Package)otherEnd, msgs);
		}
		return super.eInverseAdd(otherEnd, featureID, msgs);
	}
	
	@Override
	public NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS:
				return getBodyDeclarations().basicRemove((BodyDeclaration)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS:
				return getUsagesInTypeAccess().basicRemove((TypeAccess)otherEnd, msgs);
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE:
				return basicSetPackage(null, msgs);
		}
		return super.eInverseRemove(otherEnd, featureID, msgs);
	}
	
	public NotificationChain basicSetPackage(Package newobj, NotificationChain msgs) {
			msgs = eBasicSetContainer((InternalEObject)newobj, Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE, msgs);
			return msgs;
	}
	
	@Override
	public Object eGet(int featureID, boolean resolve, boolean coreType) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS:
				return getUsagesInTypeAccess();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS:
				return getBodyDeclarations();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_BEFORE_BODY:
				return getCommentsBeforeBody();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_AFTER_BODY:
				return getCommentsAfterBody();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE:
				return getPackage();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__SUPER_INTERFACES:
				return getSuperInterfaces();
		}
		return super.eGet(featureID, resolve, coreType);
	}
	
	@Override
	public void eSet(int featureID, Object newValue) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS:
				getUsagesInTypeAccess().clear();
				getUsagesInTypeAccess().addAll((java.util.Collection<? extends TypeAccess>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS:
				getBodyDeclarations().clear();
				getBodyDeclarations().addAll((java.util.Collection<? extends BodyDeclaration>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_BEFORE_BODY:
				getCommentsBeforeBody().clear();
				getCommentsBeforeBody().addAll((java.util.Collection<? extends Comment>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_AFTER_BODY:
				getCommentsAfterBody().clear();
				getCommentsAfterBody().addAll((java.util.Collection<? extends Comment>) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE:
				setPackage((Package) newValue);
				return;
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__SUPER_INTERFACES:
				getSuperInterfaces().clear();
				getSuperInterfaces().addAll((java.util.Collection<? extends TypeAccess>) newValue);
				return;
		}
		super.eSet(featureID, newValue);
	}
	
	@Override
	public boolean eIsSet(int featureID) {
		switch (featureID) {
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS:
				return getUsagesInTypeAccess().isEmpty(); //many, volatile
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__BODY_DECLARATIONS:
				return _bodyDeclarations != null && !_bodyDeclarations.isEmpty();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_BEFORE_BODY:
				return _commentsBeforeBody != null && !_commentsBeforeBody.isEmpty();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__COMMENTS_AFTER_BODY:
				return _commentsAfterBody != null && !_commentsAfterBody.isEmpty();
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__PACKAGE:
				return getPackage() != null; //single, volatile
			case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__SUPER_INTERFACES:
				return _superInterfaces != null && !_superInterfaces.isEmpty();
		}
		return super.eIsSet(featureID);
	}
	
	@Override
	public int eBaseStructuralFeatureID(int derivedFeatureID, Class<?> baseClass) {
		if (baseClass == Type.class) {
			switch (derivedFeatureID) {
				case Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS: return Java_PackageImpl.TYPE__USAGES_IN_TYPE_ACCESS;
				default: return -1;
			}
		}
		return super.eBaseStructuralFeatureID(derivedFeatureID, baseClass);
	}
	
				
	@Override
	public int eDerivedStructuralFeatureID(int baseFeatureID, Class<?> baseClass) {
		if (baseClass == Type.class) {
			switch (baseFeatureID) {
				case Java_PackageImpl.TYPE__USAGES_IN_TYPE_ACCESS: return Java_PackageImpl.ABSTRACT_TYPE_DECLARATION__USAGES_IN_TYPE_ACCESS;
				default: return -1;
			}
		}
		return super.eDerivedStructuralFeatureID(baseFeatureID, baseClass);
	}
	
	
}

File: crossecore-csharp/Java/Java/Java_/AbstractTypeDeclarationBase.cs
/* CrossEcore is a cross-platform modeling framework that generates C#, TypeScript, 
 * JavaScript, Swift code from Ecore models with embedded OCL (http://www.crossecore.org/).
 * The original Eclipse Modeling Framework is available at https://www.eclipse.org/modeling/emf/.
 * 
 * contributor: Simon Schwichtenberg
 */
 
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using oclstdlib;
using Ecore;
namespace Java_{
	public class AbstractTypeDeclarationBase 
	:BodyDeclarationImpl, AbstractTypeDeclaration
	{
		private Set<TypeAccess> _usagesInTypeAccess;
		
		public virtual Set<TypeAccess> usagesInTypeAccess
		{
			get {
				if(_usagesInTypeAccess==null){
					_usagesInTypeAccess = new Set<TypeAccess>(this, Java_PackageImpl.TYPE_USAGESINTYPEACCESS, Java_PackageImpl.TYPEACCESS_TYPE);
				}
				return _usagesInTypeAccess;
			}
		
		}
		private OrderedSet<BodyDeclaration> _bodyDeclarations;
		
		public virtual OrderedSet<BodyDeclaration> bodyDeclarations
		{
			get {
				if(_bodyDeclarations==null){
					_bodyDeclarations = new OrderedSet<BodyDeclaration>(this, Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS, Java_PackageImpl.BODYDECLARATION_ABSTRACTTYPEDECLARATION);
				}
				return _bodyDeclarations;
			}
		
		}
		private OrderedSet<Comment> _commentsBeforeBody;
		
		public virtual OrderedSet<Comment> commentsBeforeBody
		{
			get {
				if(_commentsBeforeBody==null){
					_commentsBeforeBody = new OrderedSet<Comment>(this, Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSBEFOREBODY, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSBEFOREBODY);
				}
				return _commentsBeforeBody;
			}
		
		}
		private OrderedSet<Comment> _commentsAfterBody;
		
		public virtual OrderedSet<Comment> commentsAfterBody
		{
			get {
				if(_commentsAfterBody==null){
					_commentsAfterBody = new OrderedSet<Comment>(this, Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSAFTERBODY, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSAFTERBODY);
				}
				return _commentsAfterBody;
			}
		
		}
		public virtual Package package
		{
			get {
			
				if (eContainerFeatureID() != Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE) return default(Package);
				return (Package)eInternalContainer();
			}
			set {
				if (value != eInternalContainer()) {
					NotificationChain msgs = null;
					if (eInternalContainer() != null){
						msgs = ((InternalEObject)eInternalContainer()).eInverseRemove(this, Java_PackageImpl.PACKAGE_OWNEDELEMENTS, typeof(AbstractTypeDeclaration), msgs);
					}
					if (value != null){
						msgs = ((InternalEObject)value).eInverseAdd(this, Java_PackageImpl.PACKAGE_OWNEDELEMENTS, typeof(AbstractTypeDeclaration), msgs);
					}
					msgs = basicSetPackage(value, msgs);
					if (msgs != null) {
						msgs.dispatch();
					}
				}
				else if (eNotificationRequired()){
					eNotify(new ENotificationImpl(this, NotificationImpl.SET,Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE , value, value));
				}
				}
		}
		private OrderedSet<TypeAccess> _superInterfaces;
		
		public virtual OrderedSet<TypeAccess> superInterfaces
		{
			get {
				if(_superInterfaces==null){
					_superInterfaces = new OrderedSet<TypeAccess>(this, Java_PackageImpl.ABSTRACTTYPEDECLARATION_SUPERINTERFACES, EOPPOSITE_FEATURE_BASE - Java_PackageImpl.ABSTRACTTYPEDECLARATION_SUPERINTERFACES);
				}
				return _superInterfaces;
			}
		
		}
		
		protected override EClass eStaticClass() {
			return Java_PackageImpl.Literals.ABSTRACTTYPEDECLARATION;
		}
		
		public override NotificationChain eInverseAdd(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE:
					if (eInternalContainer() != null) {
						msgs = eBasicRemoveFromContainer(msgs);
					}
					return basicSetPackage((Package)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS:
					return usagesInTypeAccess.basicAdd((TypeAccess)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS:
					return bodyDeclarations.basicAdd((BodyDeclaration)otherEnd, msgs);
			}
			return base.eInverseAdd(otherEnd, featureID, msgs);
		}
		
		public override NotificationChain eInverseRemove(InternalEObject otherEnd, int featureID, NotificationChain msgs) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE:
					return basicSetPackage(null, msgs);
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS:
					return usagesInTypeAccess.basicRemove((TypeAccess)otherEnd, msgs);
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS:
					return bodyDeclarations.basicRemove((BodyDeclaration)otherEnd, msgs);
			}
			return base.eInverseRemove(otherEnd, featureID, msgs);
		}
		
		public NotificationChain basicSetPackage(Package newobj, NotificationChain msgs) {
				msgs = eBasicSetContainer((InternalEObject)newobj, Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE, msgs);
				return msgs;
		}
		
		public override object eGet(int featureID, bool resolve, bool coreType) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS:
					return usagesInTypeAccess;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS:
					return bodyDeclarations;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSBEFOREBODY:
					return commentsBeforeBody;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSAFTERBODY:
					return commentsAfterBody;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE:
					return package;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_SUPERINTERFACES:
					return superInterfaces;
			}
			return base.eGet(featureID, resolve, coreType);
		}
		
		
		public override void eSet(int featureID, object newValue) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS:
					usagesInTypeAccess.Clear();
					usagesInTypeAccess.AddRange(((List<EObject>)newValue)?.Cast<TypeAccess>());
					return;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS:
					bodyDeclarations.Clear();
					bodyDeclarations.AddRange(((List<EObject>)newValue)?.Cast<BodyDeclaration>());
					return;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSBEFOREBODY:
					commentsBeforeBody.Clear();
					commentsBeforeBody.AddRange(((List<EObject>)newValue)?.Cast<Comment>());
					return;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSAFTERBODY:
					commentsAfterBody.Clear();
					commentsAfterBody.AddRange(((List<EObject>)newValue)?.Cast<Comment>());
					return;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE:
					package = (Package) newValue;
					return;
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_SUPERINTERFACES:
					superInterfaces.Clear();
					superInterfaces.AddRange(((List<EObject>)newValue)?.Cast<TypeAccess>());
					return;
			}
			base.eSet(featureID, newValue);
		}
		
		/*
		public override bool eIsSet(int featureID) {
			switch (featureID) {
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS:
					return getUsagesInTypeAccess().isEmpty(); //many, volatile
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_BODYDECLARATIONS:
					return _bodyDeclarations != null && !_bodyDeclarations.isEmpty();
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSBEFOREBODY:
					return _commentsBeforeBody != null && !_commentsBeforeBody.isEmpty();
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_COMMENTSAFTERBODY:
					return _commentsAfterBody != null && !_commentsAfterBody.isEmpty();
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_PACKAGE:
					return getPackage() != null; //single, volatile
				case Java_PackageImpl.ABSTRACTTYPEDECLARATION_SUPERINTERFACES:
					return _superInterfaces != null && !_superInterfaces.isEmpty();
			}
			return base.eIsSet(featureID);
		}
		*/
		
		public override int eBaseStructuralFeatureID(int derivedFeatureID, System.Type baseClass) {
			if (baseClass == typeof(Type)) {
				switch (derivedFeatureID) {
					case Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS: return Java_PackageImpl.TYPE_USAGESINTYPEACCESS;
					default: return -1;
				}
			}
			return base.eBaseStructuralFeatureID(derivedFeatureID, baseClass);
		}
		
					
		public override int eDerivedStructuralFeatureID(int baseFeatureID, System.Type baseClass) {
			if (baseClass == typeof(Type)) {
				switch (baseFeatureID) {
					case Java_PackageImpl.TYPE_USAGESINTYPEACCESS: return Java_PackageImpl.ABSTRACTTYPEDECLARATION_USAGESINTYPEACCESS;
					default: return -1;
				}
			}
			return base.eDerivedStructuralFeatureID(baseFeatureID, baseClass);
		}
		
		
	}
}
